### DISCUSSION TITLE
CORS origins are not working properly in my use-case

### DISCUSSION THREAD

**Discussion by AhangarAamir** [fastapi/fastapi#12171] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
## Not needed
```

### Description

I have three apps: app1, app2, and app3. In app3, I am mounting app1 and app2. app1 is restricted to certain origins, while app2 allows all origins. However, when I mount these in app3, CORS origins are not working properly. Sometimes it inherits the CORS settings from app1, and other times from app2. Note that app3 does not have any CORS middleware added.

### Operating System

Linux

### FastAPI Version

fastapi==0.111.0

### Pydantic Version

pydantic==2.7.1

### Python Version

Python 3.10.14

**Comment by JaehaSS** [comment]
You should know how the CORS middleware is working with the fastapi.
Mounted child apps are affected by the parent CORS middleware.

**Reply by AhangarAamir** [reply]
@JaehaSS in parent app i don't have any CORS middleware

**Comment by YuriiMotov** [comment]
It was probably fixed since that time, but now it seems to working as expected:

```python
from fastapi import FastAPI
from fastapi.middleware.cors import CORSMiddleware


app1 = FastAPI()
@app1.get("/")
def read_root1():
    return {"Hello": "World 1"}

app2 = FastAPI()
@app2.get("/")
def read_root2():
    return {"Hello": "World 2"}


app = FastAPI()
@app.get("/")
def read_root3():
    return {"Hello": "World 3"}


app1.add_middleware(
    CORSMiddleware,
    allow_origins=["http://localhost:3001"],
    allow_credentials=True,
    allow_methods=["*"],
    allow_headers=["*"],
)


app2.add_middleware(
    CORSMiddleware,
    allow_origins=["http://localhost:3000"],
    allow_credentials=True,
    allow_methods=["*"],
    allow_headers=["*"],
)

# Mount apps
app.mount("/app1", app1)
app.mount("/app2", app2)

```

Check:
```


curl -i -X OPTIONS http://127.0.0.1:8000/ \
  -H "Origin: http://localhost:3001" \
  -H "Access-Control-Request-Method: GET"
HTTP/1.1 405 Method Not Allowed
date: Tue, 20 May 2025 16:34:27 GMT
server: uvicorn
allow: GET
content-length: 31
content-type: application/json

{"detail":"Method Not Allowed"}



curl -i -X OPTIONS http://127.0.0.1:8000/app1/ \
  -H "Origin: http://localhost:3001" \
  -H "Access-Control-Request-Method: GET"
HTTP/1.1 200 OK
date: Tue, 20 May 2025 16:37:44 GMT
server: uvicorn
vary: Origin
access-control-allow-methods: DELETE, GET, HEAD, OPTIONS, PATCH, POST, PUT
access-control-max-age: 600
access-control-allow-credentials: true
access-control-allow-origin: http://localhost:3001
content-length: 2
content-type: text/plain; charset=utf-8


curl -i -X OPTIONS http://127.0.0.1:8000/app2/ \
  -H "Origin: http://localhost:3000" \
  -H "Access-Control-Request-Method: GET"
HTTP/1.1 200 OK
date: Tue, 20 May 2025 16:39:38 GMT
server: uvicorn
vary: Origin
access-control-allow-methods: DELETE, GET, HEAD, OPTIONS, PATCH, POST, PUT
access-control-max-age: 600
access-control-allow-credentials: true
access-control-allow-origin: http://localhost:3000
content-length: 2
content-type: text/plain; charset=utf-8
```
